runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -294,23 +300,66 @@ ILogData read(long address) {
      * @param addresses An iterable with addresses to read from
      * @return A result, which be cached.
      */
+    public Map<Long, ILogData> read(Iterable<Long> addresses) {
+        return read(addresses, true);
+    }
+
+    /**
+     * Attempts to insert a loaded value into the cache and return the cached value for a particular key.
+     */
+    private ILogData cacheLoadAndGet(Cache<Long, ILogData> cache, long address, ILogData loadedValue) {
+        try {
+            return cache.get(address, () -> loadedValue);
+        } catch (ExecutionException | UncheckedExecutionException e) {
+            // Guava wraps the exceptions thrown from the lower layers, therefore
+            // we need to unwrap them before throwing them to the upper layers that
+            // don't understand the guava exceptions
+            Throwable cause = e.getCause();
+            if (cause instanceof RuntimeException) {
+                throw (RuntimeException) cause;
+            } else {
+                throw new RuntimeException(cause);
+            }
+        }
+    }
+
+    /**
+     * Read the given object from a range of addresses.
+     *
+     * @param addresses An iterable with addresses to read from
+     * @param waitForWrite Flag whether wait for write is required or hole fill directly.
+     * @return A map of read addresses, which will be cached if caching is enabled
```

Comment:
micro nit: @return A map of addresses read, which will be cached if caching is enabled
seems to be clearer in stating that the returned values are addresses that have already been read?

Reply:
It is clearer. Changed. Thanks for the suggestion.